src/zookeeper.ts

```diff
@@ -1,316 +1,432 @@
+
+                            // further delaying the emit of 'expired' event seems to avoid
+                            // intermittent errors thrown by node-zookeeper-client
+                            setTimeout(() => {
+                                this.client.emit("expired");
+                            }, 1000);
+                        }
+                        , this.sessionTimeout);
+                });
+                this.client.on('state', (state : Object) => {
+                    this.emit(LocatorException.CODE.STATE_CHANGE, state, this.client.getSessionId());
+                });
+                this.client.on("expired", () => {
+                    this.emit(LocatorException.CODE.EXPIRED);
+                    this.connect();
+                });
+
+                this.emit(LocatorException.CODE.CONNECTING);
+                this.client.connect();
+            })
+            .catch((err) => {
+                this.emit(LocatorException.CODE.ZK_LOCATOR_ERROR, err);
+                this.connect();
+            });
     }
+}
 
-    var promises = children.map(function (child) {
-      var deferred = Promise.defer<Location>();
+/**
+ * ZookeeperPathLocatorFactory provides Locators for a list of hosts stored in a path in a zookeeper cluster
+ */
+class ZookeeperPathLocatorFactory {
+    private clientWrapper : ZookeeperClientWrapper;
+    private path : string;
+    private emitter : EventEmitter;
+    private dataExtractor : DataExtractor;
+    private locatorTimeout : number;
+    strict : boolean;
+    next : number;
+    locationPool : Location[];
+    cachedLocationPool : Location[];
+    locationRequestQueue : Promise.Resolver<Location>[];
+    stale : boolean;
+
+    /**
+     * zookeeper.client.getChildren response handler.  upon fetching the path we get a list of children nodes,
+     * call getInfo on them, and store them in the location pool and kick of the location request processing queue
+     * @param err
+     * @param children
+     */
+    private onGetLocations = (err : Error, children : string[]) => {
+        if (err) {
+            this.emitter.emit(LocatorException.CODE.FAILED_TO_GET_CHILDREN, this.path, err);
+            this.setLocationPool([]);
+            this.processQueue();
+            return;
+        }
+
+        const promises = children.map((child) => {
+            const locator = deferredLocationRequest();
 
-      clientWrapper.client.getData(path + "/" + child, (_err, data) => {
-        if (_err) {
-          if (_err.getCode() !== Exception.NO_NODE) {
-            emitter.emit(LocatorException.CODE["FAILED_TO_GET_CHILD_INFO"], path, _err);
-          }
+            this.clientWrapper.client.getData(this.path + "/" + child, (_err, data) => {
+                if (_err) {
+                    if (_err.getCode() !== zookeeper.Exception.NO_NODE) {
+                        this.emitter.emit(LocatorException.CODE.FAILED_TO_GET_CHILD_INFO, this.path, _err);
```

Comment:
I'm not entirely sure why we resolve null here instead of rejecting, like we do in other places for errors, I was just following legacy behavior. I'll do some experiments and see if it breaks anything.